packages/venia-concept/src/components/Checkout/form.js

```diff
@@ -1,329 +1,364 @@
+            });
+        },
+        [submitPaymentMethodAndBillingAddress]
+    );
+
+    const handleSubmitShippingForm = useCallback(
+        formValues => {
+            submitShippingMethod({
+                formValues
+            });
+        },
+        [submitShippingMethod]
+    );
+
+    switch (editing) {
+        case 'address': {
+            const { shippingAddress } = props;
 
-    get paymentMethodSummary() {
-        const { classes, hasPaymentMethod, paymentData } = this.props;
-
-        if (!hasPaymentMethod) {
             return (
-                <span className={classes.informationPrompt}>
-                    Add Billing Information
-                </span>
+                <AddressForm
+                    initialValues={shippingAddress}
+                    submitting={submitting}
+                    countries={countries}
+                    cancel={handleCancel}
+                    submit={handleSubmitAddressForm}
+                    isAddressIncorrect={isAddressIncorrect}
+                    incorrectAddressMessage={incorrectAddressMessage}
+                />
             );
         }
+        case 'paymentMethod': {
+            const { billingAddress } = props;
 
-        let primaryDisplay = '';
-        let secondaryDisplay = '';
-        if (paymentData) {
-            primaryDisplay = paymentData.details.cardType;
-            secondaryDisplay = paymentData.description;
+            return (
+                <PaymentsForm
+                    cancel={handleCancel}
+                    initialValues={billingAddress}
+                    submit={handleSubmitPaymentsForm}
+                    submitting={submitting}
+                    countries={countries}
+                />
+            );
         }
-
-        return (
-            <Fragment>
-                <strong className={classes.paymentDisplayPrimary}>
-                    {primaryDisplay}
-                </strong>
-                <br />
-                <span className={classes.paymentDisplaySecondary}>
-                    {secondaryDisplay}
-                </span>
-            </Fragment>
-        );
-    }
-
-    get shippingAddressSummary() {
-        const { classes, hasShippingAddress, shippingAddress } = this.props;
-
-        if (!hasShippingAddress) {
+        case 'shippingMethod': {
+            const { availableShippingMethods, shippingMethod } = props;
             return (
-                <span className={classes.informationPrompt}>
-                    Add Shipping Information
-                </span>
+                <ShippingForm
+                    availableShippingMethods={availableShippingMethods}
+                    cancel={handleCancel}
+                    shippingMethod={shippingMethod}
+                    submit={handleSubmitShippingForm}
+                    submitting={submitting}
+                />
             );
         }
+        default: {
+            return null;
+        }
+    }
+};
```

Comment:
Would be nice to have `propTypes` defined for each of these subcomponents.

Reply:
WOULDN'T IT!?



Fine :D

Reply:
And put them in new files! 💯

Reply:
DONE. 💀